equinox/nn/_selective_state_space_models.py

```diff
@@ -0,0 +1,198 @@
+import math
+from typing import Literal, Union
+
+import jax
+import jax.numpy as jnp
+from jaxtyping import Array, Float, PRNGKeyArray
+
+from .._module import field, Module
+from ._conv import Conv1d
+from ._linear import Linear
+
+
+def _selective_scan(
+    u: Float[Array, "seq_len d_inner"],
+    delta: Float[Array, "seq_len d_inner"],
+    A: Float[Array, "d_inner state_space_dims"],
+    B: Float[Array, "seq_len state_space_dims"],
+    C: Float[Array, "seq_len state_space_dims"],
+    D: Float[Array, "d_inner"],  # noqa
+):
+    seq_len, _ = u.shape
+    d_inner, state_space_dims = A.shape
+
+    delta_A = jnp.exp(jnp.einsum("l d,d n -> l d n", delta, A))
+    delta_B_u = jnp.einsum("l d,l n,l d -> l d n", delta, B, u)
+
+    x_res = jnp.zeros(shape=(d_inner, state_space_dims))
+
+    def step(x, i):
+        x = delta_A[i] * x + delta_B_u[i]
+
+        y = jnp.einsum("d n,n -> d", x, C[i, :])
+        return x, y
+
+    _, ys = jax.lax.scan(step, x_res, jnp.arange(seq_len))
```

Comment:
This should really be done with `jax.lax.associative_scan` for computing the scan in parallel. This is more faithful to the original and is much faster for long sequences on parallel processing accelerators (e.g. GPU).  Its unclear to me based on the [Griffin paper](https://arxiv.org/abs/2402.19427) if JAX's associative scan will be fater than linear scan on TPUs also, so perhaps there should be an option to choose which implementation to use.

See the S5 JAX implementation for a very similar implementation with associative scan https://github.com/lindermanlab/S5/blob/main/s5/ssm.py

Reply:
I'm not sure about this, actually. JAX's associative scan appears to completely unroll the whole operation:

https://github.com/google/jax/blob/63ceb5f539c45fe00766634ce7b01ea5176e0cc4/jax/_src/lax/control_flow/loops.py#L2171-L2186

I think there's probably a better way to implement an associative scan. Probably doing something like `equinox/internal/loop/bounded.py`, or possibly a `lax.scan` over the levels of the tree.